### DISCUSSION TITLE
Use of transit compartment models

### DISCUSSION THREAD

**Discussion by VishnuEriyat** [LAPKB/Pmetrics#214] [Q&A]
I was working on modelling Rifampicin (oral) data. The modelling of absorption seems to be particularly challenging, I wanted to explore the possibility of using transit compartment model to describe the absorption. But the run is getting halted without even starting, I tried few ways of putting the compartments, two examples are given below

@example 1
#Dif
XP(1)=-K1*X(1)-K13*X(1)
XP(2)=K1*X(1)-K23*X(2) 
XP(3)=K13*X(1)+K23*X(2)-K34*X(3)+KP43*X(4)-Ke*X(3)
XP(4)=K34*X(3)-K43*X(4)

#Out
Y(1) = X(3)/V

@example 2
#Dif
XP(1)=-K1*X(1)-Ka1*X(1)
XP(2)=K1*X(1)-Ka2*X(2) 
XP(3)=Ka1*X(1)+Ka2*X(2)-KCP*X(3)+KPC*X(4)-Ke*X(3)
XP(4)=KCP*X(3)-KPC*X(4)

#Out
Y(1) = X(3)/V

I was wondering if I am totally off the track here.

**Comment by Siel** [comment]
Hi @VishnuEriyat, thanks for reaching out.

The only thing that is weird for me in your equations is that you're missing the `*` between the different terms, although that might be a git hub formatting issue, could you paste your full model file so we can check it?

**Reply by VishnuEriyat** [reply]
[rit.txt](https://github.com/LAPKB/Pmetrics/files/14700733/rit.txt)
Thanks for the reply, The lack of * is a formatting issue.

**Comment by mnneely** [comment]
Also, your transit compartments aren’t set up correctly. 

It should be something like this: (note the current notation if using Pmetrics v2 or greater)

```
dX[1] = -Ka*X[1]
dX[2] = Ka*X[1] - Ktr*X[2]
dX[3] = Ktr*X[2] - Ktr*X[3]
dX[4] = Ktr*X[3] - (Ke + KCP)*X[4] + KPC*X[5]
dX[5] = KCP*X[4] - KPC*X[5]
```
Note that usually people make the transit rate constant the same between all transit compartments, hence Ktr between all of them.

Have you tried adding a lag time instead of transit compartments?  You can also make the lag time a function by defining the component variables in the primary block and using the LAG block to define the equation, e.g.

`TLAG[1] = function(X, y, …)`

I’ll also add that absorption can be very tricky to model sometimes, because we typically don’t have enough samples early after dosing to really characterize the curve. 

Lastly, according to this FDA [document](https://www.accessdata.fda.gov/drugsatfda_docs/label/2021/050420s086,050627s029lbl.pdf), rifampin can undergo enterohepatic circulation:

“After absorption, rifampin is rapidly eliminated in the bile, and an enterohepatic circulation
ensues. During this process, rifampin undergoes progressive deacetylation so that nearly all the
drug in the bile is in this form in about 6 hours. This metabolite has antibacterial activity.
Intestinal reabsorption is reduced by deacetylation, and elimination is facilitated. Up to 30% of a
dose is excreted in the urine, with about half of this being unchanged drug.”

If you are seeing double peaks, you should consider incorporating a delayed return to compartment 1, like this:


```
dX[1] = -Ka*X[1] + K31*X[3]
dX[2] = Ka*X[1] - K23*X[2] - (Ke + K24)*X[2] + K42*X[4]
dX[3] = K23*X[2] - K31*X[3]
dX[4] = K24*X[2] - K42*X[4]
```

**Reply by VishnuEriyat** [reply]
Thanks for pointing this out.
Yes I tried adding lag time but still many concentrations in the absorption phase is underpredicted. Defining lag time by components in the primary block is something I haven't done till now.
I am not seeing well defined second peaks, but there are profiles with increases in concentrations even after elimination sets in, shall try this delayed return and see.
Many thanks for the support

**Reply by VishnuEriyat** [reply]
I tried this model structure, but still the execution is getting halted. This is the first time Im trying a differential equation model. Will there be any compatibility issue with my Gfortran. Im gettting such a message when I start run(Screen shot attached).
![Error in execution](https://github.com/LAPKB/Pmetrics/assets/153790648/7308816a-799a-491e-8b7a-6c39aa17c3c0)

**Reply by mnneely** [reply]
@Siel , could you help here? Looks like gfortan isn't fully installed.

**Reply by Siel** [reply]
@VishnuEriyat It seems your Gfortran installation is missing some packages, how are you installing it? are you using the one Pmetrics suggests?

**Reply by VishnuEriyat** [reply]
Hi, I had installed Gfortran as Pmetrics suggested. But before I realized that Pmetrics suggests a Gfortran, I had mistakenly installed some MSys2 package also in some other way. Is it that I have 2 seperate versions of Gfortran?. But anyways my Pmetrics started running only after the suggested Gfortran was installed, it was not working with the other Msys package one.

**Reply by VishnuEriyat** [reply]
Hi any idea, what may have gone wrong here...

**Reply by mhovd** [reply]
Hi @VishnuEriyat, could you run the following command in your CMD prompt, and provide the output?

`where gfortran`

**Reply by VishnuEriyat** [reply]
![image](https://github.com/LAPKB/Pmetrics/assets/153790648/df7dcef2-7a9a-4e69-82cf-3092eca0c398)
Its coming like this.

**Reply by mhovd** [reply]
It seems you are using a version of `gfortran` provided by TDM-GCC, which I am not familiar with. Do you have Rtools install on your machine? If so, could you try to update your environment variables to use the bins in Rtools instead? On windows, you can access this by searching for "Environmental variables", then select "PATH", then click the edit button. If Rtools is not already there, press "New" and then "Browse" to locate the "bin" folder in your Rtools folder. Once you have it added, put it at the very top of your PATH, and click save. You will have to restart Rstudio, preferably your computer as well, before the changes take effect.

**Reply by VishnuEriyat** [reply]
I tried it, but did not seem to work, then finally I tried with the same TDM-gcc gfortran and added a small change in the setting(added a component 'openmp' in the current installation of TDM-gcc)). With this change the problem seems to be rectified, now execution is okay, the first cycle has started.

This idea I got from the below page
https://github.com/dmlc/xgboost/issues/1027 

By the way, Is the run times for differential equation based models very long?

**Reply by mhovd** [reply]
Solving differential equations does take longer compared to their algebraic solutions. As for _very long_, it entirely depends on your model and number of subjects/observations.

We are working on an improved backend for Pmetrics which uses Rust instead of Fortran, which based on preliminary results seems to be anywhere between two to ten times faster, depending on the model. When fully ready and integrated with Pmetrics we will be sure to post it on the discussion forum here!

**Comment by mhovd** [comment]
On the subject of absorption, there is a pretty good presentation from a PAGE meeting by Mats O. Karlsson and Rada Savic that you might find interesting.

https://www.page-meeting.org/pdf_assets/6066-karlsson_absorption.pdf

**Reply by VishnuEriyat** [reply]
Thanks, this is very useful
